File: helios/src/main/java/org/onereed/helios/sun/SunInfo.java
package org.onereed.helios.sun;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;

import java.time.Instant;

/** Information about the most recent sun event, and upcoming ones. */
@AutoValue
public abstract class SunInfo {

  public abstract Instant getTimestamp();

  public abstract float getSunAzimuthDeg();

  public abstract double getMagneticDeclinationDeg();

  public abstract int getClosestEventIndex();

  public abstract ImmutableList<SunEvent> getSunEvents();

  public static Builder builder() {
    return new AutoValue_SunInfo.Builder();
  }

  /** Builder for {@link SunInfo} */
  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setTimestamp(Instant timestamp);

    public abstract Builder setSunAzimuthDeg(float sunAzimuthDeg);

    public abstract Builder setMagneticDeclinationDeg(double magneticDeclinationDeg);

    public abstract Builder setClosestEventIndex(int closestEventIndex);

    public abstract Builder setSunEvents(ImmutableList<SunEvent> sunEvents);

    public abstract Builder setSunEvents(SunEvent... sunEvents);

    public Builder setSunAzimuthDeg(double sunAzimuthDeg) {
      return setSunAzimuthDeg((float) sunAzimuthDeg);
    }

    public abstract SunInfo build();
  }
}

File: helios/src/main/java/org/onereed/helios/sun/SunInfo.kt
package org.onereed.helios.sun

import java.time.Duration
import java.time.Instant
import timber.log.Timber

data class SunInfo(
  val sunAzimuthInfo: SunAzimuthInfo,
  val closestEventIndex: Int,
  val sunEvents: List<SunEvent>,
) {
  companion object {

    fun compute(sunTimeSeries: SunTimeSeries): SunInfo {
      Timber.d("compute start")

      val placeTime = sunTimeSeries.placeTime
      val sunEvents =
        sunTimeSeries.events.map {
          SunEvent(it.sunEventType, it.instant, placeTime.atInstant(it.instant).computeSunAzimuth())
        }

      val sunAzimuthInfo = SunAzimuthInfo.from(sunTimeSeries.placeTime)
      val closestEventIndex =
        getClosestEventIndex(sunTimeSeries.placeTime.instant, sunEvents[0], sunEvents[1])

      return SunInfo(sunAzimuthInfo, closestEventIndex, sunEvents)
    }

    private fun getClosestEventIndex(
      instant: Instant,
      lastEvent: SunEvent,
      nextEvent: SunEvent,
    ): Int =
      if (
        Duration.between(lastEvent.instant, instant) < Duration.between(instant, nextEvent.instant)
      ) {
        0
      } else {
        1
      }
  }
}
